src/emqx_router.erl

```diff
@@ -248,46 +288,36 @@ delete_trie_route(Route = #route{topic = Topic}) ->
     end.
 
 %% @private
--spec(maybe_trans(function(), list(any())) -> ok | {error, term()}).
-maybe_trans(Fun, Args) ->
+-spec(maybe_trans(pid(), function(), list(any())) -> ok | {error, term()}).
+maybe_trans(Worker, Fun, Args) ->
     case persistent_term:get(emqx_route_lock_type) of
         key ->
-            trans(Fun, Args);
+            trans(Worker, Fun, Args);
         global ->
             lock_router(),
             try mnesia:sync_dirty(Fun, Args)
             after
                 unlock_router()
             end;
         tab ->
-            trans(fun() ->
+            trans(Worker,
+                  fun() ->
                           emqx_trie:lock_tables(),
                           apply(Fun, Args)
                   end, [])
     end.
 
--spec(trans(function(), list(any())) -> ok | {error, term()}).
-trans(Fun, Args) ->
+-spec(trans(pid(), function(), list(any())) -> ok | {error, term()}).
+trans(Worker, Fun, Args) when is_pid(Worker) ->
     %% trigger selective receive optimization of compiler,
     %% ideal for handling bursty traffic.
-    Ref = erlang:make_ref(),
-    Owner = self(),
-    {WPid, RefMon} = spawn_monitor(
-                       fun() ->
-                               Res = case mnesia:transaction(Fun, Args) of
-                                         {atomic, Ok} -> Ok;
-                                         {aborted, Reason} -> {error, Reason}
-                                     end,
-                               Owner ! {Ref, Res}
-                       end),
+    MonRef = erlang:monitor(process, Worker),
```

Comment:
if spawn_link, there is no need to monitor.

Reply:
spawn_link is bidirectional which is not ideal at this point.